File: backend/online/server/dto/conversation.go
package dto

import (
	"encoding/json"
	"time"
)

type CreateConversationRequest struct {
	Novel      string    `json:"novel"`
	ShortStory string    `json:"short_story"`
	Poem       string    `json:"poem"`
	Play       string    `json:"play"`
	Film       string    `json:"film"`
	By         string    `json:"by"`
	Rule       string    `json:"rule"`
	Capacity   int       `json:"capacity"`
	When       time.Time `json:"when"`
	Length     string    `json:"length"`
}

type ConversationFeedResponse struct {
	Id           string    `json:"id"`
	Novel        string    `json:"novel,omitempty"`
	ShortStory   string    `json:"shortStory,omitempty"`
	Poem         string    `json:"poem,omitempty"`
	Play         string    `json:"play,omitempty"`
	Film         string    `json:"film,omitempty"`
	By           string    `json:"by,omitempty"`
	Rule         string    `json:"rule,omitempty"`
	When         time.Time `json:"when"`
	Length       string    `json:"length"`
	Ongoing      bool      `json:"ongoing"`
	IsModerator  bool      `json:"isModerator"`
	IsRegistrant bool      `json:"isRegistrant"`
}

type ConversationSignalResponse struct {
	FromIds []string        `json:"fromIds"`
	Signal  json.RawMessage `json:"signal,omitempty"`
}

type ConversationSignalRequest struct {
	ToIds  []string        `json:"toIds"`
	Signal json.RawMessage `json:"signal"`
}

type ConversationSignalMessage struct {
	FromId string          `json:"fromId"`
	ToId   string          `json:"toId"`
	Signal json.RawMessage `json:"signal,omitempty"`
}

type GetConversationResponse struct {
	Id         string    `json:"id"`
	Novel      string    `json:"novel,omitempty"`
	ShortStory string    `json:"shortStory,omitempty"`
	Poem       string    `json:"poem,omitempty"`
	Play       string    `json:"play,omitempty"`
	Film       string    `json:"film,omitempty"`
	By         string    `json:"by,omitempty"`
	Rule       string    `json:"rule,omitempty"`
	When       time.Time `json:"when"`
	Length     string    `json:"length"`

	IsModerator bool `json:"isModerator"`
}

File: client/mobile/types/conversation.ts
export interface CreateOnlineConversationRequest {
  novel?: string;
  shortStory?: string;
  poem?: string;
  play?: string;
  film?: string;
  by?: string;
  rule?: string;
  capacity: number;
  when: string;
  length: string;
}

export interface OnlineConversationFeedResponse {
  id: string;
  novel?: string;
  shortStory?: string;
  poem?: string;
  play?: string;
  film?: string;
  by?: string;
  rule?: string;
  capacity: number;
  when: string;
  length: string;
  ongoing: boolean;
  isModerator: boolean;
  isRegistrant: boolean;
}

export interface BanParticipantRequest {
  conversationId: string;
  banId: string;
}

export interface ConversationSignalResponse {
  fromIds: string[];
  signal?: PeerSignal;
}

type PeerSignal =
  | { type: "offer" | "answer"; sdp: string }
  | { type: "candidate"; candidate: RTCIceCandidate }
  | { type: "leave" }
  | { type: "ban" }
  | { type: "mute" }
  | { type: "name-offer" | "name-answer"; name: string };

interface RTCIceCandidate {
  candidate?: string;
  sdpMLineIndex?: number | null;
  sdpMid?: string | null;
}

export type SeatCoordinate = { left: number; top: number };


export interface GeoCoordinates {
  lat: number;
  lng: number;
}

export interface SeatAssignment {
  id?: string;
  name?: string;
  mute?: boolean;
  left: number;
  top: number;
}

export interface OfflineConversationMapResponse {
  id: string;
  writtenBy: string;
  lat: number;
  lng: number;
}

export interface CreateOfflineConversationRequest {
  novel?: string;
  shortStory?: string;
  poem?: string;
  play?: string;
  film?: string;
  writtenBy: string;
  rule?: string;
  time: string;
  length: number;
  googleMapsLink: string;
  location: string;
  city?: string;
  lat: number;
  lng: number;
  h3Res5: string;
  h3Res7: string;
}
